resource/Resource.md

```diff
@@ -0,0 +1,154 @@
+application and passing it to a backend-specific exporter. This explicitly includes
+the path through future OpenCensus components such as agents or services.
+
+For example, process-identifying information may be populated through the library while
+an agent attaches further labels about the underlying VM, the cluster, or geo-location.
+
+### From environment variables
+Population of resource information from environment variables MUST be provided by the
+core library. It provides the user with an ubiquitious way to manually provide information
+that may not be detectable automatically through available integration libraries.
+
+Two environment variables are used:
+* `OC_RESOURCE_TYPE`: defines the resource type. Leading and trailing whitespaces are trimmed.
+* `OC_RESOURCE_LABELS`: defines resource labels as a comma-seperated list of key/value pairs.
+The key MUST be seperated from the value by a single `=`. Values MAY be quoted with a single
+leading and trailing `"`. If a values contains whitespaces, `=`, or `"` characters it MUST be
+quoted and `"` characters MUST be escaped.
+
+For example:
+* `OC_RESOURCE_TYPE=k8s.io/container`
+* `OC_RESOURCE_LABELS=k8s.io/pod_name="pod-xyz-123",k8s.io/container_name="c1",k8s.io/namespace="default"`
+
+Population from environment variables MUST be the first applied detection process unless
+the user explicitly overwrites this behavior.
+
+### Auto-detection
+Auto-detection of resource information in specific environments, e.g. specific cloud
+vendors, MUST be implemented outside of the core libraries in third party or
+[census-ecosystem][census-ecosystem] repositories.
+
+### Merging
+As different mechanisms are run to gain information about a resource, their information
+has to be merged into a single resulting resource.
+Already set labels or type fields MUST NOT be overwritten. Label key namespaceing SHOULD
+be used to prevent collisions across different resource detection steps.
+
+### Detectors
+To make auto-detection implementations easy to use, the core resource package SHOULD define
+an interface to retrieve resource information. Additionally, helper functionality MAY be
+provided to effectively make use of this interface.
+The exact shape of those interfaces and helpers SHOULD be idiomatic to the respective language.
+
+Example in Go:
+
+```go
+type Detector func(context.Context) (*Resource, error)
+
+// Returns a detector that always returns a specific resource.
+func NewDetectorFromResource(*Resource) Detector
+
+// Returns a detector that runs all input detectors sequentially and merges their results.
+func ChainedDetector(...Detector) Detector
+```
+
+### Updates
+Over the runtime of an application, resource information may change in some cases. To
+allow exporters to incorporate those changes, they SHOULD take a resource getter as a
+configuration parameter. The exporter SHOULD periodically rerun the getter and incorporate
```

Comment:
should there be notification mechanism for subscribers on updates? It will be useful to have if you have an exporter that needs to react on these changes. Otherwise there will be a mess of threads checking for updates from each other

Reply:
Right now I think a periodic rerun of the detector and comparing with the previous result to detect changes will be sufficient. But there may very well be other use cases I didn't think of.

Note that the exporter-specific part of this spec is largely _SHOULD_, so there's enough flexibility to deviate/extend where necessary.

Reply:
heatlth and lifetime management are scenarios that might require more immidiate update of properties.

Reply:
@SergeyKanzhelev I would like to understand the "heatlth and lifetime management are scenarios". Can you give me more details about what you have in mind.

My personal opinion is to keep simplicity and actually not support runtime config for these in the library at all, unless we have a clear use-case which I miss here, but I can be convinced if you have one in mind.

Reply:
The key of heatlth and lifetime management is that different modules may contribute into the common state and make decisions based on it. Like if CPU is too high - make sampling more aggressive and report yellow state of the app on this node. Or if exporter failed to upload telemetry with many retries - switch an app to red state and potentially node out of rotation. Or when app was swapped into production from staging - start collecting more metrics. 

I agree with you that resource concept like proposed can be implemented if there is a way to set global attributes like proposed in https://github.com/census-instrumentation/opencensus-specs/pull/167. Maybe make it more elaborate and implement it via callbacks to resource auto-discovery modules (without calling them that).

So static and almost static informaiton is not that hard to implement without introducing new concepts (except global attributes or attributes callbacks).